<a href="https://colab.research.google.com/github/tirabo/Algoritmos-y-Programacion/blob/main/Maximo_comun_divisor_numeros_perfectos_Clase_laboratorio_03_05_21.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Máximo común divisor

Veamos como calcular el máximo común divisor de dos números "inspeccionando" los divisores comunes. Es decir si $a$, $b$  enteros, uno de ellos no negativo 
$$\operatorname{mcd}(a,b) = \text{el máximo divisor común a $a$ y $b$.}
$$  


In [ ]:
# primero definimos una función que nos devuelve los divisores de un número
def divisores(a):
  return  [d for d in range(1, a+1) if a % d == 0]

# veamos ahora varias formas de calcular el mcd, todas ellas parecidas, 
# siempre eligiendo el divisor común más grande. 

def maximo_comun_divisor_v1(a, b: int) -> int:
  # pre: (a > 0) or (b > 0)
  # post:  devuelve mcd(a,b)
  div_de_a = divisores(a)
  div_de_b = divisores(b)
  i = len(div_de_a)
  j = len(div_de_b)
  mcd = False # indica que no se ha encontrado el mcd
  while not mcd:
    k = j
    for _ in range(j):
      if div_de_a[i - 1] == div_de_b[k - 1] :
        div = div_de_a[i - 1]
        mcd = True
      k -= 1
    i -= 1
  return div

def maximo_comun_divisor_v2(a, b: int) -> int:
  # pre: (a > 0) or (b > 0)
  # post:  devuelve mcd(a,b)
  div_de_a = divisores(a)
  div_de_b = divisores(b)
  h = len(div_de_a) - 1
  mcd = False # indica que no se ha encontrado el mcd
  while not mcd:
    if div_de_a[h] in div_de_b: # usa la relación 'in'
      div = div_de_a[h]
      mcd = True
    h = h - 1
  return div

def maximo_comun_divisor_v3(a, b: int) -> int:
  # pre: (a > 0) or (b > 0)
  # post:  devuelve mcd(a,b)
  div_de_a = divisores(a)
  div_de_b = divisores(b)
  divisores_comunes = [(x,y) for x in div_de_a for y in div_de_b if x == y]
  # la lista por comprensión anterior está formada por todos los pares (x,x) 
  # donde x es divisor de a y x es divisor de b
  return max(divisores_comunes)[0]

# Una versión optimizada de este algoritmo es la siguiente
def maximo_comun_divisor_v4(a, b: int) -> int:
  # pre: (a > 0) or (b > 0)
  # post:  devuelve mcd(a,b)
  div_de_a, div_de_b = divisores(a), divisores(b)
  # las listas anteriores están ordenadas pero por las dudas, y para no depender
  # de la implementación de otra función, las ordenamos.
  div_de_a.sort(), div_de_b.sort()
  i, j = len(div_de_a) - 1, len(div_de_b) - 1
  while div_de_a[i] != div_de_b[j]:
    if div_de_a[i] < div_de_b[j]:
      j = j - 1
    else:
      i = i - 1
  return div_de_a[i]

# En realidad la versión optimizada no aporta mucho, pues lo más costoso a 
# nivel computacional es la función divisores() y las listas de divisores no son 
# demasiado grandes (en longitud).  



print(maximo_comun_divisor_v4(12, 15))
print(maximo_comun_divisor_v4(10000, 23456))
print(maximo_comun_divisor_v4(10000, 110))
print(maximo_comun_divisor_v4(3*2**20, 3**10*2))

In [ ]:
# Otro ejemplo de definición de lista por comprensión
# Dadas dos lista hace la lista "intersección". 
lista1 = divisores(1000)
lista2 = divisores(55)
elemento = [x for x in lista1 if x in lista2]
print(lista1)
print(lista2)
print(elemento)

# Python también permite trabajar con conjuntos
conjunto1, conjunto2 = set(lista1), set(lista2)
print(conjunto1)
print(conjunto2)
interseccion = conjunto1.intersection(conjunto2)
print(interseccion)



## Números perfectos

Un número perfecto es un número entero positivo que es igual a la suma de sus divisores propios positivos. 

Por ejemplo, $6 = 3 + 2 +1$.

*Ejercicio.* Calcular todos los números perfectos  menores e iguales que $10000$.


In [ ]:
def perfectos(n: int) -> list:
  # post:  devuelve la lista de enteros perfectos hasta n
  perfectos = []
  for i in range(1, n + 1):
    div_i = divisores(i)
    div_i.sort() # ordenamos div_i
    suma = 0
    for k in range(len(div_i) - 1):
      suma = suma + div_i[k]
    # suma = sumatoria de todos los divisores propios de i
    if suma == i:
      perfectos.append(i)
  return perfectos

print(perfectos(10000))

## Números perfectos pares

**Teorema de Euclides-Euler**

*Un número par es perfecto si y sólo si tiene la forma $2^{p−1}(2^p - 1)$, donde $2^p - 1$ es un número primo.*   

In [ ]:
def es_primo(n):
    # pre: n entero positivo
    # post:  devuelve True si es primo, si no devuelve False
    div_n = divisores(n)
    return len(div_n) == 2 # n es primo <=> tiene exactamente dos divisores positivos 

def numeros_perfectos_pares(n):
    # pre: n entero positivo
    # post: devuelve la lista de numeros perfectos pares <= n
    perfectos = []
    p = 1
    candidato =  2**(p-1)*(2**p -1)
    while candidato <= n:
        #print(candidato,(2**p -1), [d for d in range(1, 2**p) if n % d == 0]) )
        if es_primo(2**p-1):
            perfectos.append(candidato)
        p = p + 1
        candidato = 2**(p-1)*(2**p -1)
    return perfectos

# Este algoritmo nos permite encontrar más números perfectos que el anterior.
print(numeros_perfectos_pares(10**12))



[6, 28, 496, 8128, 33550336, 8589869056, 137438691328]


El último primo perfecto $\le 10^{12}$ es 
$$137438691328 = 2^{18}(2^{19} -1).$$
El próximo es $2^{30}(2^{31} -1)$ pero no podemos hallarlo con nuestro método. La dificultad reside en que la función `divisores()` es muy poco eficiente  y por lo tanto la verificación `es_primo()` tampoco es eficiente. 

Existes algoritmos muy eficientes para determinar si un número es primo o no. Estos algoritmos no se basan en encontrar los divisores, si no en ciertas propiedades algebraicas que poseen los números primos. Con el uso de estos algoritmos,  quizás en alguna clase hablemos de ellos, se puede avanzar más en la búsqueda de números perfectos. 

Por último, mencionemos (ver Wikipedia) que a la fecha:
- Se han encontrado $47$ números perfectos, todos pares.
- No se sabe si existen números perfectos impares.
- No se sabe si hay infinitos números perfectos.

El  número perfecto  más grande que se conoce fue hallado en el año 2009 y es
$$
2^{43112608}(2^{ 43112609} -1).
$$
Este número tiene casi $13$ millones de dígitos. 